Improve Windows qmake build configuration files

FranciscoPombal opened the pull request:
This comes from my recent experience with https://github.com/qbittorrent/qBittorrent/issues/12357.

- ~Given that Windows is what it is, seems useless to support anything other than static builds there (this is the main reason this is a draft PR, it is a big change in something I'm not totally comfortable with)~
- The names of the crypto libs have changed; I also changed the name of the boost lib to not hardcode a specific version.
- ~We don't officially support any platforms that don't have OpenSSL >= 1.1 for some time now (when support for RC_1_1 is removed the other nearby `TODO` can be addressed).~

NotTsunami commented:
Are we sure that TODO is correct for the OpenSSL define? That define seems to still be used quite frequently, especially in the new [HTTPS tracker validation](https://github.com/arvidn/libtorrent/commit/4fd6136b2ab8a45517a06e621805408b0d611cdd#diff-7d96b2f4364fd1f465ab84e94a770554).

jagannatharjun commented:
Non static builds are useful for developers🙃

Chocobo1 commented:
conf.pri.windows is supposed to be customized by each user: https://github.com/qbittorrent/qBittorrent/blob/master/winconf.pri#L56
IMO the current version in master is not outdated nor need any improvements suggested here.

glassez commented:
>Non static builds are useful for developers

Non static builds are usedul not only for developers! The fact that qBittorrent uses static linking for official builds is mostly personal preference of project maintainer.
>Given that Windows is what it is, seems useless to support anything other than static builds there 

Please don't try to pass off your personal opinion as something undeniable. You risk losing your trust, IMO...

FranciscoPombal commented:
@glassez @jagannatharjun 

> > Given that Windows is what it is, **seems** useless to support anything other than static builds there
> 
> Please don't try to pass off your personal opinion as something undeniable. You risk losing your trust, IMO...

> **seems**

Not trying to pass anything as undeniable... This was just my understanding. It can be wrong and I'm happy to be corrected if that is the case.

> > Non static builds are useful for developers
> 
> Non static builds are useful not only for developers! The fact that qBittorrent uses static linking for official builds is mostly personal preference of project maintainer.

Can you explain why/how it can be useful, please? I'm legitimately curious and want to learn (especially now that I'm dealing more with Windows builds due to the nightly builds PR).

@NotTsunami 

> Are we sure that TODO is correct for the OpenSSL define?

I'll admit I just trusted that TODO at face value. Maybe the TODO itself should be removed then?

@Chocobo1

At least the crypto library name change would be useful right? Like I said, we don't support anything that does not have OpenSSL >= 1.1. I don't see a reason to keep supporting lower OpenSSL versions if higher ones are supported everywhere else, but please let me know if there are reasons to do so.

> LIBS += zlib.lib libcrypto.lib libssl.lib

sledgehammer999 commented:
>We don't officially support any platforms that don't have OpenSSL >= 1.1 for some time now

AFAIK we never had a requirement for specific version of OpenSSL. We don't explicitly depend on it. We do that indirectly through Qt and libtorrent. So we don't care about the version used, as long as, qt/libtorrent supports it.

As @Chocobo1 said, the file is there as a template.

The only changes that **may** be accepted are the lib names for OpenSSL. But imo, there should be a commented out line with the lib names of openssl 0.9.x (?)

FranciscoPombal commented:
@sledgehammer999 

> AFAIK we never had a requirement for specific version of OpenSSL. We don't explicitly depend on it. We do that indirectly through Qt and libtorrent. So we don't care about the version used, as long as, qt/libtorrent supports it.
> 
> As @Chocobo1 said, the file is there as a template.
> 
> The only changes that **may** be accepted are the lib names for OpenSSL. But imo, there should be a commented out line with the lib names of openssl 0.9.x (?)

IMO templates should be updated every now and then. Additionally, they should allow and _encourage_ a modern and standard configuration out-of-the-box.

Chocobo1 commented:
>Like I said, we don't support anything that does not have OpenSSL >= 1.1. I don't see a reason to keep supporting lower OpenSSL versions if higher ones are supported everywhere else, but please let me know if there are reasons to do so.

I've said it many times, we only use what we need. Other lib uses whatever version they want and we don't care. Quit complicating things!

>AFAIK we never had a requirement for specific version of OpenSSL. We don't explicitly depend on it. 

Actually we do. It is for password hashing in utils/password.cpp and so we do have a version requirement written in INSTALL file.

>The only changes that may be accepted are the lib names for OpenSSL. 

I think this is acceptable.

sledgehammer999 commented:
>Actually we do. It is for password hashing in utils/password.cpp and so we do have a version requirement written in INSTALL file.

oh! I missed that.
Then the lib names for OpenSSL in `conf.pri.win` must be changed to what @FranciscoPombal has done. IIRC they are the same in openssl 1.0.x too.

FranciscoPombal commented:
@sledgehammer999 @Chocobo1
Addressed comments. Basically the structure of the file is a little different now and the crypto library names are changed.

Another thing. Our boost version requirement is at 1.40. That's really old. Couldn't we bump it to something like 1.65.1 (version on oldest Ubuntu version that we support)? That would also enable us to remove some stuff in the build files related to old boost versions.

Chocobo1 commented on conf.pri.windows:
```diff
@@ -13,17 +13,18 @@ LIBS += $$quote(-LC:/qBittorrent/openssl/lib)
 # Adapt the lib names/versions accordingly
 # If you want to use Boost auto-linking then disable
 # BOOST_ALL_NO_LIB below and omit Boost libraries here
+LIBS += zlib.lib libcrypto.lib libssl.lib
```
please move it back, it is intended being there.

Chocobo1 replied:
Or we can maintain the order by lib names in each section, i.e. with the following order: [boost, libtorrent, openssl, zlib, qbt specifics].

Neumann-A replied:
ZLIB also carries a debug suffix on windows (-> zlibd.lib)

FranciscoPombal replied:
Done

Chocobo1 commented on conf.pri.windows:
```diff
@@ -13,17 +13,18 @@ LIBS += $$quote(-LC:/qBittorrent/openssl/lib)
-LIBS += zlib.lib
+
 # ...or if you use MinGW
+#LIBS += libz libcrypto libssl
```
this too.

FranciscoPombal replied:
@Chocobo1
I noticed both branches of the conditional of the mingw config are the same. So is the suffix the same in both configurations? or are the debug suffixes missing? Please let me know so I can fix it.

Chocobo1 replied:
I won't always have answer to your questions, it would be better to make the change *after* you found an answer.

FranciscoPombal replied:
@Chocobo1 I'm only asking you because I figured you would probably know, since you have used mingw in other contexts: https://github.com/Chocobo1/qbittorent_msys2. I personally don't use it and don't really care about it, so I'd rather not have to test it myself if others already know the answer.

Chocobo1 commented:
>That would also enable us to remove some stuff in the build files related to old boost versions.

Could you please point out where.

FranciscoPombal commented:
@Chocobo1

> > That would also enable us to remove some stuff in the build files related to old boost versions.
> 
> Could you please point out where.

Here:
https://github.com/qbittorrent/qBittorrent/blob/26941a5b02a4281ca3608dd8cca69b6531306e41/conf.pri.windows#L43
and here:
https://github.com/qbittorrent/qBittorrent/blob/26941a5b02a4281ca3608dd8cca69b6531306e41/winconf.pri#L1

There may be more, I did not perform an exhaustive search.

Chocobo1 commented:
> > > That would also enable us to remove some stuff in the build files related to old boost versions.
> > 
> > Could you please point out where.
> 
> Here:
> 
> https://github.com/qbittorrent/qBittorrent/blob/26941a5b02a4281ca3608dd8cca69b6531306e41/conf.pri.windows#L43

Maybe it can bump to boost 1.60.


> and here:
> 
> https://github.com/qbittorrent/qBittorrent/blob/26941a5b02a4281ca3608dd8cca69b6531306e41/winconf.pri#L1

You'll need confirmation from sledgehammer999 on this.

FranciscoPombal commented:
> Maybe it can bump to boost 1.60.

Why not 1.65.1, as mentioned in https://github.com/qbittorrent/qBittorrent/pull/12418#issuecomment-609486250?

> > and here:
> > https://github.com/qbittorrent/qBittorrent/blob/26941a5b02a4281ca3608dd8cca69b6531306e41/winconf.pri#L1
> 
> You'll need confirmation from sledgehammer999 on removing this.

@sledgehammer999 ping, is that workaround still needed in recent versions? Have you been building official releases with it?

Chocobo1 commented:
> > Maybe it can bump to boost 1.60.
> 
> Why not 1.65.1, as mentioned in [#12418 (comment)](https://github.com/qbittorrent/qBittorrent/pull/12418#issuecomment-609486250)?

You only need 1.60 to get rid of it, it is irrational to make decisions that way.

FranciscoPombal commented:
@glassez Isn't the current set of changes still controversial? Why hand on to the old name of `ssleay32`, for example?

glassez commented:
>@glassez Isn't the current set of changes still controversial? Why hand on to the old name of ssleay32, for example?

How about "d" suffix you added for all libraries? I'm not sure it's used with any of them except the Boost one.

FranciscoPombal commented:
@glassez 

> How about "d" suffix you added for all libraries? I'm not sure it's used with any of them except the Boost one.

I believe you are aware that in this context, `libxyzd` means "`libxyz` compiled with debugging symbols". Doesn't it make sense to default to the versions of the libraries compiled with debugging symbols in a debug build?

You may say that these names may not "just work" due to not coinciding with what the user is using by default. But historically, this file never "just worked as is", because these names were always simply guidelines that the user would most likely have to change to fit their exact use-case (the previous `libboost_system-vc140-mt-sgd-1_64.lib` is an example of this, as the version is hard-coded).

In light of this, I think having the `d` names is a good default, because it is suggestive of what one might want to do (for example, it is very convenient to have `libtorrent` debugging symbols in qBittorrent debug builds).

glassez commented:
>I believe you are aware that in this context, libxyzd means "libxyz compiled with debugging symbols". 

**d** suffix is only used in some (rare) cases to speculate about it here...
>You may say that these names may not "just work" due to not coinciding with what the user is using by default. But historically, this file never "just worked as is"

Yes. But I would prefer it not to have deliberately false examples in it.

I'm sorry, your argument seems subjective and unconvincing to me.
The only thing that could be changed is the names of OpenSSL libraries according to the new naming in v1.1. Everything else does not deserve to be merged, IMO.
But I'm more neutral about it. So if someone else supports these changes, you can continue without my explicit approval.

FranciscoPombal commented:
@glassez to be honest, I don't really mind that much not using the `d` prefix for the others enough for this to be delayed any longer. I'll revert that change, but I should keep the `d` in boost, correct?

glassez commented:
>but I should keep the d in boost, correct?

Only Boost use this prefix so it can be used here as well. Other libraries have the same name for each build type so we should have only one entry per each other library.

FranciscoPombal commented:
@glassez fixed and rebased.

glassez commented:
Thank you.
